src/core/xccl_lib.c

```diff
@@ -0,0 +1,120 @@
+#include <xccl_lib.h>
+#include <ucs/config/types.h>
+#include <ucs/debug/log_def.h>
+#include <ucs/config/parser.h>
+#include "utils/xccl_log.h"
+#include <xccl_team_lib.h>
+
+extern xccl_lib_t xccl_static_lib;
+
+static ucs_config_field_t xccl_lib_config_table[] = {
+
+   {NULL}
+};
+UCS_CONFIG_REGISTER_TABLE(xccl_lib_config_table, "XCCL", NULL, xccl_lib_config_t)
+
+#define CHECK_LIB_CONFIG_CAP(_cap, _CAP_FIELD) do{                        \
+        if ((params->field_mask & XCCL_LIB_PARAM_FIELD_ ## _CAP_FIELD) && \
+            !(params-> _cap & tl->params. _cap)) {                        \
+            printf("Disqualifying team %s due to %s cap\n",               \
+                   tl->name, UCS_PP_QUOTE(_CAP_FIELD));                   \
+            continue;                                                     \
+        }                                                                 \
+    } while(0)
+
+
+static void xccl_lib_filter(const xccl_lib_params_t *params, xccl_lib_t *lib)
+{
+    int i;
+    int n_libs = xccl_static_lib.n_libs_opened;
+    lib->libs = (xccl_team_lib_t**)malloc(sizeof(xccl_team_lib_t*)*n_libs);
+    lib->n_libs_opened = 0;
+    for (i=0; i<n_libs; i++) {
+        xccl_team_lib_t *tl = xccl_static_lib.libs[i];
+        CHECK_LIB_CONFIG_CAP(reproducible, REPRODUCIBLE);
+        CHECK_LIB_CONFIG_CAP(thread_mode,  THREAD_MODE);
+        CHECK_LIB_CONFIG_CAP(team_usage,   TEAM_USAGE);
+        CHECK_LIB_CONFIG_CAP(coll_types,   COLL_TYPES);
+        
+        lib->libs[lib->n_libs_opened++] = tl;
+    }
+}
+
+xccl_status_t xccl_lib_init(const xccl_lib_params_t *params,
+                            const xccl_lib_config_t *config,
+                            xccl_lib_h *xccl_lib)
+{
+    xccl_lib_t *lib;
+
+    if (xccl_static_lib.n_libs_opened == 0) {
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    lib = malloc(sizeof(*lib));
+    if (lib == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+
+    xccl_lib_filter(params, lib);
+    if (lib->n_libs_opened == 0) {
+        xccl_error("XCCL lib init: no plugins left after filtering by params\n");
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    *xccl_lib = lib;
+    //TODO: move to appropriate place
+    //ucs_config_parser_warn_unused_env_vars_once("XCCL_");
+    return XCCL_OK;
+}
+
+xccl_status_t xccl_lib_config_read(const char *env_prefix,
+                                   const char *filename,
+                                   xccl_lib_config_t **config_p){
+    xccl_lib_config_t *config;
+    xccl_status_t status;
+    char full_prefix[128] = "XCCL_";
+
+    config = malloc(sizeof(*config));
```

Comment:
@Sergei-Lebedev we need to decide how we process malloc return - it should be same everywhere. E.g., above in the diff where ctx config is malloced - there is not check for OOM. If we check it - we should do it everywhere.